### ISSUE TITLE
Certificate chain is incomplete on macOS

### ISSUE THREAD

**Issue by KINFOO** [PeculiarVentures/fortify#537]
Hi

# Context

I am using `fortify-webcomponents-react` a `yubikey` to generate a signature.
I also retrieve certificate chain with this code

```jsx
export const SignatureView = (/* stuff */) => {
  const onSelectionSuccess = async (event: any) => {
    try {
      const provider = await event.detail.socketProvider.getCrypto(event.detail.providerId);
      provider.sign = provider.subtle.sign.bind(provider.subtle);

      const certificate = await provider.certStorage.getItem(event.detail.certificateId);

      setEngine(
        'newEngine',
        new CryptoEngine({
          name: nanoid(),
          crypto: provider,
          subtle: provider.subtle,
        })
      );

      const chain = (await provider.certStorage.getChain(certificate)).reverse();
      alert(chain.length);
    } catch (error) {
      onError('Fortify is unable to communicate with smartcard.');
      console.error(error);
    }
  };
  return (
    <PeculiarFortifyCertificates
      hide-footer
      filters={{ onlySmartcards: true, onlyWithPrivateKey: true, keyUsage: ['digitalSignature'] }}
      onSelectionSuccess={onSelectionSuccess}
      onSelectionCancel={() => onCancel()}
    />
  );
};
```

# Problem

Certificate chain in incomplete on _macOS_!

## Look

Running the code above on __Linux__

<p align="center"><img src="https://user-images.githubusercontent.com/154818/220586089-c890ee53-fbcc-47ae-82e2-b5a92a06a500.png"/></p>

Running the code above on __macOS__

<p align="center"><img width="527" alt="mac" src="https://user-images.githubusercontent.com/154818/220586923-c6ed2a11-92c3-41c3-93de-acb41088f164.png"></p>

# Expected output

Certificate chain length should be the same on all operating systems

# Version

## Yubikey

Yubikey 5 NFC
Firmware: `5.4.3`

This Yubikey obviously manages [signing using your example](https://codepen.io/donskov/pen/OJMPPNX)

<p align="center"><img width="818" alt="yubikey" src="https://user-images.githubusercontent.com/154818/220588112-4408f3a0-0ea5-4b68-a157-5c937d973367.png"></p>

## fortify-webcomponents-react

`3.0.0`

## Linux

```sh
$ uname -a
Linux 5.10.0-20-amd64 #1 SMP Debian 5.10.158-2 (2022-12-13) x86_64 GNU/Linux
```

Fortify: `1.8.3`

## macOS

Processor: _Apple M2 Pro_
macOS: `13.2.1`
Fortify: `1.8.4`

🔈 Installation process notified me Fortify has been installed using [Rosetta](https://en.wikipedia.org/wiki/Rosetta_(software)) 🔈

**Comment by rmhrisk** [comment]
As the Fortify code is the same across platforms this has to be one of the following:

1. A difference in the Yubikey middleware behavior
2. A difference in the Yubikey provisioning m
3. A difference in the OSX configuration 
4. A difference in the OSX behavior 

If so, there is not a lot that can be done besides possibly adding a layer that does a chain build and fetches intermediates by chasing the AIA. Does the certificate in question have a valid AIA:issuer field?

@microshine am I missing any other variants? Can you think of anything we can do to help here?

**Comment by microshine** [comment]
I think the most impossible reason on it is

> A difference in the Yubikey middleware behavior

Do you use the latest version of PIV Yubico tools?

If it works fine on Linux it must work on MacOS too. App uses PKIjs module to build chain from certificates on the token.

Here is our test of `getChain` function. Just tried it on my MacOS machine, it works fine with SoftHSM.

https://github.com/PeculiarVentures/webcrypto-local/blob/master/test/client/cert_storage.ts#L163

I've got a Yubico token but I can't import certificates into it using PIV tools. @KINFOO Do you know how to import certificates into the Yubico token?

**Comment by microshine** [comment]
Please try to run this script with your token. This script outputs a list of certificates on the token.

```js
import { Crypto } from "node-webcrypto-p11";
import * as x509 from "@peculiar/x509";

async function main() {
  const crypto = new Crypto({
    library: "/usr/local/lib/libykcs11.dylib",
    slot: 0,
    pin: "<your pin here>",
  });

  try {
    const keys = await crypto.certStorage.keys();
    for (const key of keys) {
      if (!key.startsWith("x509-")) {
        continue;
      }
      const raw = await crypto.certStorage.getValue(key);
      const cert = new x509.X509Certificate(raw);
      console.log("Certificate:", key);
      console.log("  Subject:", cert.subject);
      console.log("  Issuer:", cert.issuer);
    }
  } finally {
    crypto.close();
  }
}

main().catch(e => {
  console.error(e);
  process.exit(1);
})
```

### Output

```
Certificate: x509-2600000000000000-02
  Subject: CN=microshine
  Issuer: CN=microshine
```

**Comment by KINFOO** [comment]
Hi,

@rmhrisk On _Linux_ and _macOS_ I used the same Yubikey.
I mean the very same physical Yubikey

@microshine Thanks for the details

I noticed `libykcs11.dylib` is available at:

```
/opt/homebrew/lib/libykcs11.dylib
/opt/homebrew/Cellar/yubico-piv-tool/2.3.1/lib/libykcs11.dylib
```

So I updated your script, see it running:

```js
import { Crypto } from "node-webcrypto-p11";
import * as x509 from "@peculiar/x509";

async function main() {
  const crypto = new Crypto({
    library: "/opt/homebrew/lib/libykcs11.dylib",
    slot: 0,
    pin: "<3",
  });

  try {
    const keys = await crypto.certStorage.keys();
    for (const key of keys) {
      if (!key.startsWith("x509-")) {
        continue;
      }
      const raw = await crypto.certStorage.getValue(key);
      const cert = new x509.X509Certificate(raw);
      console.log("Certificate:", key);
      console.log("  Subject:", cert.subject);
      console.log("  Issuer:", cert.issuer);
    }
  } finally {
    crypto.close();
  }
}

main().catch(e => {
  console.error(e);
  process.exit(1);
})
```

Running the code I get:

```
$ node --version
v16.19.0
$ node main.js
Certificate: x509-2500000000000000-01
  Subject: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudSigner, E=signer@yellowcloud.org
  Issuer: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudInter2, E=inter2@yellowcloud.org
Certificate: x509-2900000000000000-05
  Subject: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudInter2, E=inter2@yellowcloud.org
  Issuer: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudInter1, E=inter1@yellowcloud.org
Certificate: x509-2a00000000000000-06
  Subject: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudInter1, E=inter1@yellowcloud.org
  Issuer: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudCA, E=ca@yellowcloud.org
Certificate: x509-2b00000000000000-07
  Subject: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudCA, E=ca@yellowcloud.org
  Issuer: C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudCA, E=ca@yellowcloud.org
Certificate: x509-3d00000000000000-19
  Subject: CN=Yubico PIV Attestation
  Issuer: CN=Yubico PIV Root CA Serial 263751
```

I checked the path of my middleware at `.fortify/cards.json`,

I see:

```json
    {
      "id": "993988460d8f49a2ac519a2935f00533",
      "name": "YubiKey",
      "file": {
        "osx": "/usr/local/lib/libykcs11.dylib",
        "linux": "/usr/local/lib/libykcs11.so",
        "windows": "%WINDIR/System32/libykcs11-1.dll"
      }
    },
```

Changing it to:

```json
    {
      "id": "993988460d8f49a2ac519a2935f00533",
      "name": "YubiKey",
      "file": {
        "osx": "",
        "linux": "/usr/local/lib/libykcs11.so",
        "windows": "%WINDIR/System32/libykcs11-1.dll"
      }
    },
```

I reboot _Fortify_ deamon...

Same issue on webcomponents

<p align="center"><img width="529" alt="Capture d’écran 2023-02-23 à 18 19 11" src="https://user-images.githubusercontent.com/154818/220982057-2b77939c-5dd1-4d42-ab42-92a988b179f5.png"></p>

**Comment by microshine** [comment]
@KINFOO, could you please update your script for `fortify-webcomponents-react`? Does it return the same number of certificate identifiers as the `node-webcrypto-p11` script? If so, there seems to be an issue with the implementation of the `getChain` method in the Fortify application. If the number of identifiers is lower, it would be interesting to know which PKCS#11 library Fortify is using to access the token. This can be seen through the log file ~/.fortify/fortify.log.

### fortify-webcomponents-react

```js
const provider = await event.detail.socketProvider.getCrypto(event.detail.providerId);

// get all ids from the certStorage
const keys = await provider.certStorage.keys();
for (const key of keys) {
  if (!key.startsWith("x509-")) {
    continue;
  }
  console.log(key);
}
```

### log example

```json
{"atr":"3bfc1300008131fe15597562696b65794e454f7233e1","level":"info","message":"New token was added to the reader","reader":"Yubico Yubikey NEO OTP+U2F+CCID","source":"pcsc","timestamp":"2023-02-15T13:39:56.821Z"}
{"atr":"3bfc1300008131fe15597562696b65794e454f7233e1","level":"info","message":"Token was added to the reader","name":"Yubico Yubikey NEO OTP+U2F+CCID","reader":"Yubico Yubikey NEO OTP+U2F+CCID","source":"provider","timestamp":"2023-02-15T13:39:58.118Z"}
{"level":"info","library":"/usr/local/lib/libykcs11.dylib","message":"Loading PKCS#11 library","source":"provider","timestamp":"2023-02-15T13:39:58.119Z"}
{"level":"info","message":"Looking for slot","slots":1,"source":"provider","timestamp":"2023-02-15T13:39:59.422Z"}
{"level":"info","message":"Use ConfigTemplateBuilder","source":"provider","timestamp":"2023-02-15T13:40:00.051Z"}
{"cryptokiVersion":{"major":2,"minor":40},"firmwareVersion":{"major":1,"minor":0},"level":"info","library":"/usr/local/lib/libykcs11.dylib","libraryVersion":{"major":2,"minor":30},"manufacturerId":"Yubico (www.yubico.com)","message":"PKCS#11 library information","source":"provider","timestamp":"2023-02-15T13:40:00.052Z"}
{"id":"e96f0bd16bf92e3b4f1f6139ed6bc858bdb70eec6716530e2038c6ea17d4b5aa","level":"info","library":"/usr/local/lib/libykcs11.dylib","message":"Crypto provider was added to the list","name":"Yubico Yubikey NEO OTP+U2F+CCID","reader":"Yubico Yubikey NEO OTP+U2F+CCID","source":"provider","timestamp":"2023-02-15T13:40:00.053Z"}
```

**Comment by microshine** [comment]
You can enable logging using Fortify preferences 

- Click `Preferences...` in Fortify tray menu
- Enable `Logging`

**Comment by microshine** [comment]
If Fortify cannot load PKCS#11 library specified in `card.json` it tries to get access to the token using OpenSC library which is embedded in Fortify application. Looks like Fortify uses another lib than `/opt/homebrew/lib/libykcs11.dylib` on your Mac. Log file should help us to check out this.

**Comment by rmhrisk** [comment]
@KINFOO can we see that log file?

**Comment by KINFOO** [comment]
Hi

Here is my `~/.fortify/fortify.log` 

```json
{"level":"info","message":"Starting","source":"application","time":"2023-03-07T10:22:32.523Z","timestamp":"2023-03-07T10:22:32.534Z"}
{"level":"info","message":"Application info","source":"application","timestamp":"2023-03-07T10:22:32.590Z","version":"1.8.4","versionChrome":"91.0.4472.164","versionElectron":"13.6.9","versionNode":"14.16.0"}
{"arch":"x64","cpus":12,"freemem":865710080,"level":"info","message":"System info","platform":"darwin","source":"system","timestamp":"2023-03-07T10:22:32.598Z","totalmem":34359738368,"type":"Darwin","version":"13.2.1"}
{"height":982,"level":"info","message":"Screen size","source":"system","timestamp":"2023-03-07T10:22:32.808Z","width":1512}
{"lang":"en","level":"info","message":"Change language","source":"l10n","timestamp":"2023-03-07T10:22:32.810Z"}
{"level":"info","message":"Get certificate status","source":"ssl-service","status":"valid","timestamp":"2023-03-07T10:22:32.842Z"}
{"command":"\"/Applications/Fortify.app/Contents/MacOS/certutil\" -L -d \"sql:/Users/kevin/Library/Application Support/Firefox/Profiles/7lyvd8xo.default\" -h \"all\"","level":"info","message":"Run certutil command","shell":"bash","source":"nss","timestamp":"2023-03-07T10:22:32.845Z"}
{"error":"Command failed: \"/Applications/Fortify.app/Contents/MacOS/certutil\" -L -d \"sql:/Users/kevin/Library/Application Support/Firefox/Profiles/7lyvd8xo.default\" -h \"all\"\ncertutil: function failed: SEC_ERROR_BAD_DATABASE: security library: bad database.\n","level":"error","message":"SSL install error","source":"ssl-installer","stack":"Error: Command failed: \"/Applications/Fortify.app/Contents/MacOS/certutil\" -L -d \"sql:/Users/kevin/Library/Application Support/Firefox/Profiles/7lyvd8xo.default\" -h \"all\"\ncertutil: function failed: SEC_ERROR_BAD_DATABASE: security library: bad database.\n\n    at checkExecSyncError (child_process.js:625:11)\n    at execSync (child_process.js:661:15)\n    at electron/js2c/asar_bundle.js:5:12288\n    at t.NssCertUtils.run (/Applications/Fortify.app/Contents/Resources/app/out/main.js:329:186279)\n    at t.NssCertUtils.list (/Applications/Fortify.app/Contents/Resources/app/out/main.js:329:185715)\n    at t.NssCertUtils.exists (/Applications/Fortify.app/Contents/Resources/app/out/main.js:329:185456)\n    at t.SslCertInstaller.installFirefox (/Applications/Fortify.app/Contents/Resources/app/out/main.js:329:183506)\n    at g.run (/Applications/Fortify.app/Contents/Resources/app/out/main.js:329:178972)\n    at t.Server.init (/Applications/Fortify.app/Contents/Resources/app/out/main.js:103:168129)\n    at b.initServer (/Applications/Fortify.app/Contents/Resources/app/out/main.js:244:90441)","timestamp":"2023-03-07T10:22:32.978Z"}
{"command":"\"/Applications/Fortify.app/Contents/MacOS/certutil\" -L -d \"sql:/Users/kevin/Library/Application Support/Firefox/Profiles/t9s9exzz.default-release\" -h \"all\"","level":"info","message":"Run certutil command","shell":"bash","source":"nss","timestamp":"2023-03-07T10:22:32.978Z"}
{"command":"\"/Applications/Fortify.app/Contents/MacOS/certutil\" -L -d \"sql:/Users/kevin/Library/Application Support/Firefox/Profiles/t9s9exzz.default-release\" -n \"Fortify Local CA\" -a \"\"","level":"info","message":"Run certutil command","shell":"bash","source":"nss","timestamp":"2023-03-07T10:22:33.072Z"}
{"level":"info","message":"SSL certificate is loaded","source":"server","timestamp":"2023-03-07T10:22:33.165Z"}
{"level":"info","message":"Comparing current version of card.json file with remote","source":"server","timestamp":"2023-03-07T10:22:33.167Z"}
{"level":"info","message":"card.json has the latest version","source":"server","timestamp":"2023-03-07T10:22:33.393Z","version":"1.1.11"}
{"level":"info","message":"Start PCSC listening","source":"pcsc","timestamp":"2023-03-07T10:22:33.557Z"}
{"level":"info","message":"Check for new update","source":"update","timestamp":"2023-03-07T10:22:33.575Z"}
{"duration":1056,"level":"info","message":"Loaded","source":"application","time":"2023-03-07T10:22:33.579Z","timestamp":"2023-03-07T10:22:33.580Z"}
{"address":"127.0.0.1:31337","level":"info","message":"Started","source":"server","timestamp":"2023-03-07T10:22:33.586Z"}
{"cryptokiVersion":{"major":2,"minor":30},"firmwareVersion":{"major":0,"minor":1},"level":"info","library":"/Applications/Fortify.app/Contents/Resources/app/libpvpkcs11.dylib","libraryVersion":{"major":1,"minor":2},"manufacturerId":"Module","message":"PKCS#11 library information","source":"provider","timestamp":"2023-03-07T10:22:33.634Z"}
{"id":"53311cf0ccd7be304171d05a366e06e3e30968a4","level":"info","library":"/Applications/Fortify.app/Contents/Resources/app/libpvpkcs11.dylib","message":"Crypto provider was added to the list","name":"MacOS Crypto","reader":"MacOS Crypto","source":"provider","timestamp":"2023-03-07T10:22:33.636Z"}
{"level":"info","message":"Use default TemplateBuilder","source":"provider","timestamp":"2023-03-07T10:22:33.643Z"}
{"cryptokiVersion":{"major":2,"minor":20},"firmwareVersion":{"major":1,"minor":0},"level":"info","library":"/Applications/Fortify.app/Contents/MacOS/libsoftokn3.dylib","libraryVersion":{"major":3,"minor":37},"manufacturerId":"Mozilla Foundation","message":"PKCS#11 library information","source":"provider","timestamp":"2023-03-07T10:22:33.644Z"}
{"id":"8bbae6f98340d7a92006c21da65c1acd345ba946","level":"info","library":"/Applications/Fortify.app/Contents/MacOS/libsoftokn3.dylib","message":"Crypto provider was added to the list","name":"NSS Certificate DB","reader":"NSS User Private Key and Certificate Services","source":"provider","timestamp":"2023-03-07T10:22:33.645Z"}
{"level":"info","message":"Start PCSC listening","source":"pcsc","timestamp":"2023-03-07T10:22:33.649Z"}
{"level":"info","message":"New version wasn't found","source":"update","timestamp":"2023-03-07T10:22:33.852Z"}
{"level":"info","message":"Create a new connection","origin":"https://qa-kinfoo.airvantage.io","source":"server","timestamp":"2023-03-07T10:23:20.563Z"}
{"level":"info","message":"Push session to stack","origin":"https://qa-kinfoo.airvantage.io","source":"server","timestamp":"2023-03-07T10:23:20.565Z"}
{"level":"warn","message":"Cannot parse MessageSignedProtocol","source":"server","timestamp":"2023-03-07T10:23:20.591Z"}
{"authorized":true,"level":"info","message":"Initialize secure session","origin":"https://qa-kinfoo.airvantage.io","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:20.617Z"}
{"action":"server/isLoggedIn","level":"info","message":"Run action","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:20.623Z"}
{"action":"provider/action/info","level":"info","message":"Run action","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:20.644Z"}
{"level":"info","message":"Initialize new reader","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"pcsc","timestamp":"2023-03-07T10:23:25.798Z"}
{"atr":"3bfd1300008131fe158073c021c057597562694b657940","level":"info","message":"New token was added to the reader","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"pcsc","timestamp":"2023-03-07T10:23:25.804Z"}
{"level":"info","message":"Initialize new reader","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"pcsc","timestamp":"2023-03-07T10:23:25.875Z"}
{"atr":"3bfd1300008131fe158073c021c057597562694b657940","level":"info","message":"New token was added to the reader","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"pcsc","timestamp":"2023-03-07T10:23:25.879Z"}
{"atr":"3bfd1300008131fe158073c021c057597562694b657940","level":"info","message":"cardReader/insert","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"server-api","timestamp":"2023-03-07T10:23:26.808Z"}
{"atr":"3bfd1300008131fe158073c021c057597562694b657940","level":"info","message":"Token was added to the reader","name":"Yubico Yubikey 4 OTP+U2F+CCID","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"provider","timestamp":"2023-03-07T10:23:26.933Z"}
{"level":"info","library":"/opt/homebrew/lib/libykcs11.dylib","message":"Loading PKCS#11 library","source":"provider","timestamp":"2023-03-07T10:23:26.935Z"}
{"error":"dlopen(/opt/homebrew/lib/libykcs11.dylib, 0x0006): tried: '/opt/homebrew/lib/libykcs11.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libykcs11.dylib' (no such file), '/opt/homebrew/lib/libykcs11.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/opt/homebrew/Cellar/yubico-piv-tool/2.3.1/lib/libykcs11.2.3.1.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/Cellar/yubico-piv-tool/2.3.1/lib/libykcs11.2.3.1.dylib' (no such file), '/opt/homebrew/Cellar/yubico-piv-tool/2.3.1/lib/libykcs11.2.3.1.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64'))","level":"error","message":"Server event error","source":"server","timestamp":"2023-03-07T10:23:26.940Z"}
{"level":"info","library":"/Applications/Fortify.app/Contents/MacOS/opensc-pkcs11.so","message":"Loading PKCS#11 library","source":"provider","timestamp":"2023-03-07T10:23:26.943Z"}
{"level":"info","message":"Looking for slot","slots":1,"source":"provider","timestamp":"2023-03-07T10:23:27.960Z"}
{"level":"info","message":"Use default TemplateBuilder","source":"provider","timestamp":"2023-03-07T10:23:27.988Z"}
{"cryptokiVersion":{"major":2,"minor":20},"firmwareVersion":{"major":0,"minor":0},"level":"info","library":"/Applications/Fortify.app/Contents/MacOS/opensc-pkcs11.so","libraryVersion":{"major":0,"minor":20},"manufacturerId":"OpenSC Project","message":"PKCS#11 library information","source":"provider","timestamp":"2023-03-07T10:23:27.999Z"}
{"id":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","level":"info","library":"/Applications/Fortify.app/Contents/MacOS/opensc-pkcs11.so","message":"Crypto provider was added to the list","name":"Yubico Yubikey 4 OTP+U2F+CCID","reader":"Yubico YubiKey OTP+FIDO+CCID","source":"provider","timestamp":"2023-03-07T10:23:28.001Z"}
{"added":1,"level":"info","message":"Amount of tokens was changed","removed":0,"source":"provider-service","timestamp":"2023-03-07T10:23:28.002Z"}
{"action":"provider/action/info","level":"info","message":"Run action","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:28.049Z"}
{"action":"provider/action/getCrypto","level":"info","message":"Run action","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:28.083Z"}
{"action":"crypto/isLoggedIn","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:28.112Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"crypto/isLoggedIn","source":"server-api","timestamp":"2023-03-07T10:23:28.113Z"}
{"action":"crypto/login","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:28.129Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"crypto/login","source":"server-api","timestamp":"2023-03-07T10:23:28.130Z"}
{"level":"info","message":"Create window","name":"p11-pin","source":"windows","timestamp":"2023-03-07T10:23:28.229Z"}
{"level":"info","message":"Close window","name":"p11-pin","source":"windows","timestamp":"2023-03-07T10:23:35.800Z"}
{"action":"crypto/keyStorage/keys","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:35.874Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"keyStorage/keys","source":"server-api","timestamp":"2023-03-07T10:23:35.886Z"}
{"action":"crypto/certificateStorage/keys","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:35.905Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/keys","source":"server-api","timestamp":"2023-03-07T10:23:35.907Z"}
{"action":"crypto/certificateStorage/getItem","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:35.922Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","index":"x509-2041640000600000-01","level":"info","message":"certStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:35.929Z"}
{"cert":{"publicKey":{"algorithm":{"label":"PIV AUTH pubkey","name":"ECDSA","namedCurve":"P-256","sensitive":false,"token":true},"extractable":true,"id":"01","type":"public","usages":["verify"]},"subjectName":"C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudSigner, E=signer@yellowcloud.org","token":true,"type":"x509"},"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:35.967Z"}
{"action":"crypto/certificateStorage/export","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:35.994Z"}
{"cert":{"publicKey":{"algorithm":{"label":"PIV AUTH pubkey","name":"ECDSA","namedCurve":"P-256","sensitive":false,"token":true},"extractable":true,"id":"01","type":"public","usages":["verify"]},"subjectName":"C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudSigner, E=signer@yellowcloud.org","token":true,"type":"x509"},"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/exportCert","source":"server-api","timestamp":"2023-03-07T10:23:35.997Z"}
{"action":"provider/action/getCrypto","level":"info","message":"Run action","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:40.674Z"}
{"action":"crypto/certificateStorage/keys","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:40.714Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/keys","source":"server-api","timestamp":"2023-03-07T10:23:40.718Z"}
{"action":"crypto/certificateStorage/getItem","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:40.741Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","index":"x509-2041640000600000-01","level":"info","message":"certStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:40.742Z"}
{"cert":{"publicKey":{"algorithm":{"label":"PIV AUTH pubkey","name":"ECDSA","namedCurve":"P-256","sensitive":false,"token":true},"extractable":true,"id":"01","type":"public","usages":["verify"]},"subjectName":"C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudSigner, E=signer@yellowcloud.org","token":true,"type":"x509"},"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:40.765Z"}
{"action":"crypto/keyStorage/getItem","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:40.801Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","index":"private-0040640000600000-01","level":"info","message":"keyStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:40.804Z"}
{"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","key":{"algorithm":{"label":"PIV AUTH key","name":"ECDSA","namedCurve":"P-256","sensitive":true,"token":true},"extractable":false,"id":"01","type":"private","usages":["sign"]},"level":"info","message":"keyStorage/getItem","source":"server-api","timestamp":"2023-03-07T10:23:40.805Z"}
{"action":"crypto/certificateStorage/getChain","level":"info","message":"Run action","provider":"b16ca252e8404535a826bc598bfa670a4078da6ae6c599e8dbd03b2e8759d7a7","session":"0156d1302804c998679e64be3259770e181854dac07b283212c9597736b20b14","source":"server","timestamp":"2023-03-07T10:23:40.825Z"}
{"cert":{"publicKey":{"algorithm":{"label":"PIV AUTH pubkey","name":"ECDSA","namedCurve":"P-256","sensitive":false,"token":true},"extractable":true,"id":"01","type":"public","usages":["verify"]},"subjectName":"C=US, ST=CA, L=SF, O=YellowCloud, OU=DataBall, CN=YellowCloudSigner, E=signer@yellowcloud.org","token":true,"type":"x509"},"crypto":"Yubico Yubikey 4 OTP+U2F+CCID","level":"info","message":"certStorage/chain","source":"server-api","timestamp":"2023-03-07T10:23:40.829Z"}
{"level":"info","message":"CKA_X509_CHAIN is not supported","source":"server-api","timestamp":"2023-03-07T10:23:40.841Z"}
```

Well, looks like I need `libykcs11.dylib` for `x86_64`...

```
{"error":"dlopen(/opt/homebrew/lib/libykcs11.dylib, 0x0006): tried: '/opt/homebrew/lib/libykcs11.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/System/Volumes/Preboot/Cryptexes/OS/opt/homebrew/lib/libykcs11.dylib' (no such file), '/opt/homebrew/lib/libykcs11.dylib' (mach-o file, but is an incompatible architecture (have 'arm64', need 'x86_64')), '/opt/homebrew/Cellar/yubico-piv-tool/2.3.1
```

So I managed to install [`libykcs11.dylib` for `x86_64`](https://medium.com/mkdir-awesome/how-to-install-x86-64-homebrew-packages-on-apple-m1-macbook-54ba295230f)

I now have it at

```sh
$ file /Users/kevin/.local/homebrew/lib/libykcs11.dylib
/Users/kevin/.local/homebrew/lib/libykcs11.dylib: Mach-O 64-bit dynamically linked shared library x86_64
```

So I update `~/.fortify/card.json`

```json
    {
      "id": "993988460d8f49a2ac519a2935f00533",
      "name": "YubiKey",
      "file": {
        "osx": "/users/kevin/.local/homebrew/lib/libykcs11.dylib",
        "linux": "/usr/local/lib/libykcs11.so",
        "windows": "%WINDIR/System32/libykcs11-1.dll"
      }
    },
```

And 💥 BOOM 💥 it works

<p align="center"><img width="514" alt="Capture d’écran 2023-03-07 à 22 47 58" src="https://user-images.githubusercontent.com/154818/223562210-2d10a6cb-7f36-4dfd-a62d-16a287924550.png"></p>

## Long story short

If you install **Fortify** _x86_64_ using [Rosetta](https://en.wikipedia.org/wiki/Rosetta_(software)) , you need _x86_64_ `.dylib` in your `~/.fortify/card.json`


Thanks for the steps @microshine & @rmhrisk
